### ISSUE TITLE
Regular expression globbing documentation lacks important details

### ISSUE THREAD

**Issue by inventhouse** [xonsh/xonsh#3372]
I'm new to xonsh and very excited to have things like `re` at my fingertips at all times, but my first attempt to use re-globbing produced a strange and confusing error message:
```
# find .py files 0-or-more directories deep
$ r`(.*/)*\w+\.py`                                                                          
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
re.error: missing ), unterminated subpattern at position 0
```

Clearly, I thought, my subpattern is balanced, and nothing in the [documentation] lead me to believe that it would be parsed any differently than standard `re`:

> Any string that uses backticks (`) instead of quotes (', ") is interpreted as a regular expression to match filenames against.

[documentation]: https://xon.sh/tutorial.html#regular-expression-globbing

I was able to find [the code], which showed me that the expression is split and matched segment-by-segment, which is not necessarily a bad way to do it, but does create very important limitations and gotchas that should be documented.  It would also be helpful if the standard error message thrown by `re` were augmented to include the pattern (or partial-pattern) and a bit of explanation.

[the code]: https://xon.sh/_modules/xonsh/built_ins.html#reglob

**Comment by inventhouse** [comment]
You can, of course, easily do a full-depth search and filter by a complete pattern if you want:

`[ p for p in g`**/*` if re.match(r"(.*/)*\w+\.py", p) ]`

**Comment by inventhouse** [comment]
For posterity:
```
$ xonsh --version                                                                           
xonsh/0.9.13
```

**Comment by gforsyth** [comment]
Thanks for reporting, @inventhouse ! We'd welcome any clarifying documentation (or we can look into changing the implementation to better match expectations).

**Comment by astronouth7303** [comment]
Ummm... that's a bug. That's a valid regular expression. You should not be getting that error.

**Comment by inventhouse** [comment]
I don't mind that the implementation is more like "path-wise regular expression", as that can prune a great deal of the search space down in mosts cases; also the workaround to do a full search by a complete pattern is trivial, whereas to re-implement the more efficient "path-wise" search is not.

I just think it needs to be documented and perhaps a small improvement to the error messages; when I have a bit of time, I think this would be a good first contribution.

**Comment by inventhouse** [comment]
Although...  I just realized that if we dig into the `re` guts enough, we might be able to test a partial path against a full pattern and determine if the pattern consumes all of the string; if it does, continue to search that branch, but if it is actively rejected, prune.  That could get us similar efficiency without butchering the pattern.

**Comment by inventhouse** [comment]
Yes, I linked to the code in the description and summarized what it's doing; "mangling" is a reasonable adjective 😜 [Here's the link again](https://xon.sh/_modules/xonsh/built_ins.html#reglob)

**Comment by astronouth7303** [comment]
... I think the real solution to that is to attempt to match windows paths based on slashes and backslashes, instead of trying to mangle the regex.

**Comment by inventhouse** [comment]
Posted a question to StackOverflow for this, if anyone can point me in the right direction, you could save me a bunch of time implementing this: https://stackoverflow.com/questions/58599729/programmatically-determine-where-pythons-re-match-stopped-matching

**Comment by inventhouse** [comment]
As far as I can tell, the standard `re` module hides the specific reason a match fails in [_sre_SRE_Pattern_match_impl](https://github.com/python/cpython/blob/d0e0f5bf0c07ca025f54df21fd1df55ee430d9fc/Modules/_sre.c#L589), inaccessible (as far as I know) from Python, so my algorithm above can't be done without modifying `re` itself or finding a pure-python implementation.

I split out fixing this "right" as #3381 but, in the meantime, this issue can continue to track documenting the current behavior and improving the error messages.

**Comment by inventhouse** [comment]
Not sure why this one fails (though I haven't really looked hard, just making a note before I forget)

> ⫸ \`(\w+)/(\1)\.py\`
> re.error: cannot refer to an open group at position 8
> ⫸ [ p for p in g\`**/*.py\` if re.match(r"(\w+)/(\1)\.py", p) ]
> ['statemachine/statemachine.py']
